Source/WebCore/loader/NavigationScheduler.cpp

```diff
@@ -359,6 +360,19 @@ class ScheduledHistoryNavigationByKey : public ScheduledNavigation {
         }
 
         auto completionHandler = std::exchange(m_completionHandler, nullptr);
+        auto rawEntries = page->checkedBackForward()->allItems();
+        for (size_t i = 0; i < rawEntries.size(); i++) {
+            Ref item = rawEntries[i];
+            auto index = item->children().findIf([&historyItem](const auto& entry) {
+                return entry == historyItem;
+            });
+            if (index != notFound) {
+                if (item->documentSequenceNumber() != page->backForward().currentItem()->documentSequenceNumber()) {
+                    historyItem = item;
```

Comment:
In my patch I made a clone of the tree instead of modifying it in place which *seems* more correct.

Reply:
This is required for many tests so it seems good for now.

Reply:
> This is required for many tests so it seems good for now. There is actually a trade off because a handful work if it is a clone:
> 
> * traverseTo-with-cross-origin-in-history
> * navigation-methods/traverseTo-with-cross-origin-in-history
>   (few more)

Yes it is quite possible both our PRs have some strong points that maybe can be merged, I'll have a closer look at your PR tomorrow.